pkg/apis/mpi/validation/validation.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2018 The Kubeflow Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validation
+
+import (
+	"fmt"
+
+	commonv1 "github.com/kubeflow/common/pkg/apis/common/v1"
+
+	mpiv1 "github.com/kubeflow/training-operator/pkg/apis/mpi/v1"
+)
+
+func ValidateV1MpiJobSpec(c *mpiv1.MPIJobSpec) error {
+	if c.MPIReplicaSpecs == nil {
+		return fmt.Errorf("MPIReplicaSpecs is not valid")
+	}
+	launcherExists := false
+	for rType, value := range c.MPIReplicaSpecs {
+		if value == nil || len(value.Template.Spec.Containers) == 0 {
+			return fmt.Errorf("MPIReplicaSpecs is not valid: containers definition expected in %v", rType)
```

Comment:
Generally, when validating APIs, you should return all possible errors. See https://github.com/kubeflow/mpi-operator/blob/master/v2/pkg/apis/kubeflow/validation/validation.go